python/testSrc/com/jetbrains/python/PyTypeTest.java

```diff
@@ -1702,6 +1702,31 @@ public void testMaxListUnknownElements() {
            "expr = max(l)");
   }
 
+  public void testWithAsType() {
+    doTest("ClassA",
+           "class ClassA(object):\n" +
+           "    def __enter__(self):\n" +
+           "        return self;\n" +
+           "\n" +
+           "    def __exit__(self, exc_type, exc_val, exc_tb):\n" +
+           "        return\n" +
+           "\n" +
+           "class ClassB(object):\n" +
+           "    def __init__(self):\n" +
+           "        super(ClassB, self).__init__()\n" +
+           "        self.a = None\n" +
+           "\n" +
+           "    def fun(self):\n" +
+           "        if self.a is not None:\n" +
+           "            return self.a\n" +
+           "        self.a = ClassA()\n" +
+           "        return self.a\n" +
+           "\n" +
+           "foo=ClassB()\n" +
+           "with foo.fun() as expr:\n" +
```

Comment:
Please simplify your test by leaving all the non-essential bits. For example, the current version tests not only using expressions of union types in `with`-statements, but type inference for instance attributes assigned in methods (`self.a = ClassA()`), type inference for `None` checks (`if self.a is not None`) and so on.

I would prefer something along these lines:

```python
from typing import Union

class A(object):
    def __enter__(self):
        return self

class B(object):
    def __enter__(self):
        return self

def f(x):
    # type: (Union[A, B]) -> None
    with x as expr:
        pass
```